gitpod-io/gitpod
# [ws-daemon] Support proper PID namespace-relative /proc mounts

This PR let's supervisor mount `/proc` in a workspace container, within the correct PID namespace.
This is a prerequisite for properly supporting Docker (see #2096) and the PID namespace in ring2.

### What's the problem?
`/proc` provides access to un-namespaced objects, e.g. `/proc/kcore` or `/proc/sched_debug`. From a security perspective those objects are a problem, e.g. `/proc/sched_debug` leaks information about other namespaces. To mitigate those security issues, `/proc` [is commonly masked](https://github.com/kubernetes/kubernetes/blob/master/pkg/securitycontext/util.go#L200-L218). Once `/proc` is masked however, one cannot simply mount `/proc` elsewhere as that would render the whole mask useless. More precisely, if you can't see a full `/proc`, you can't mount `/proc` either ([`mount_too_revealing`](https://github.com/torvalds/linux/blob/0eac1102e94807023e57d032bbba51830928b78e/fs/namespace.c#L2802)).

We need to mount `/proc` for two use-cases:
1. we want to wrap `ring2` in its own PID namespace, to prevent direct access to the daemon socket and mark mount. To this end we need to give the ring2 rootfs its own `/proc` mount, mounted from the PID namespace of `ring2`.
2. we need proper `/proc` mounts in the Docker containers our users will start (see #2096)

Note that simply disabling the proc mask in Kubernetes (`procMount: Unmasked`) does not help. This would solve the PID namespace issue, but wouldn't work for the Docker container, because the "workspace" must not have access to an unmasked `/proc`. Also, it would weaken the security of the workspace container, should someone find a way out of `ring2`.

### How does it work?
We make `ws-daemon` do the `/proc` mount on our behalf. To this end, the `InWorkspaceService` gets a new API call: `MountProc`. This MountProc call does the following:
1. map the in-container PID of the process we want to mount `/proc` for to the host PID space
2. `nsenter` the target PID namespace, but stay in ws-daemon's mount namespace
3. mount proc to a temporary folder within ws-daemon
4. mount the masks on top of that proc mount
5. move the mount (`mount --move`) to the workspace's daemon hostPath mount (same place we put the mark mount or daemon socket)
6. `nsenter` the target mount namespace and move the proc mount in place

This process assumes that the daemon hostPath mount `/.workspace` is visible from the target namespace. 
Because the mask mounts are made "off-site", i.e. outside of the target mount namespace, the user cannot unmount them, and there is no point in time where the proc mount would be unmasked and accessible from the workspace.

Now that we can provide proper PID-namespace sensitive `/proc` mounts, we can use this ability to isolate `ring2` in a PID namespace. To this end, ring2 gets started in its own PID namespace, ring1 mounts `/proc` using ws-daemon and signals ring2 that things are ready. ring2 then `pivot_root`s into the new rootfs prepared in ring1. By unmounting the old root (i.e. the root prior to pivot_root) we make `ring2` loose all connections to `ring1`, hence completely isolate the user's workload from the supervisor rings/mounts/socket.

The following diagram outlines this process:
![image](https://user-images.githubusercontent.com/3210701/97865799-117a3700-1d0b-11eb-86b1-f78458b7e460.png)


### How to test?
1. Log in to http://cw-ws-daemon-proc-mount.staging.gitpod-dev.com/workspaces/ to make sure your user exists
2. Enable permissions/feature flags by running `leeway run components/ws-daemon:enabled-userns-ff`
3. Start a workspace
4. Inspect `/proc`, `/proc/1/cmdline`, `/proc/1/root` which should point to `supervisor ring2` and not show anything you can't find in `/` anyways
5. run `top` - the "root process" should be supervisor ring2
6. run `mount` and see if you find anything suspicious
7. compare `ls -lh /proc/1/ns` with `/proc/self/ns`, they should be the same, i.e. your process and the "root process" should be in the same namespaces
8. see if you can unmount anything that's masked in `/proc`
9. see if you can access anything in `/proc` that you [shouldn't](https://github.com/gitpod-io/gitpod/blob/0c7daacb0e397bb7edfb713f743069920900042f/components/ws-daemon/pkg/iws/iws.go#L53-L70)

### Future work
1. after this PR user-namespaced workspaces are safe enough to no longer require the `privileged-ws` permission. We should do away with that requirement.
2. sometimes userns workspaces don't stop properly because their final call to `ws-daemon` failes. That's not a problem created by this PR, but still one we need to solve.
3. integrate with  the OCI hooks introduced in #2096. To do this we need to expose some `mountProc` API towards the workspace.
4. do the same we've done for `/proc` for `/sys` because Docker runs in its own network namespace. For `ring2` we're fine to just bind mount the workspace container's `/sys`.
